### ISSUE TITLE
Include specific Dicom group on get_identifiers

### ISSUE THREAD

**Issue by W7CAM** [pydicom/deid#71]
Is there a way to ensure that specific tag groups are included? Currently, the get_identifer function does not retrieve the 0051 group (which is essential for my image reconstruction...) Thanks!

**Comment by vsoch** [comment]
hey @W7CAM ! I would recommend that you copy the default deid specification file, and then write the exact criteria that you need. The file provided here is a simple (base case) template as it would of course be challenging for it to fit everyone's needs! Notice that the get_identifiers function (in dicom/identifiers.py) has a config argument, by default it's False to use the default provided:

```
def get_identifiers(dicom_files,
                    force=True,
                    config=None,
                    expand_sequences=True,
                    skip_fields=None):
...
```
The default provided has a set called "skip" when it's loading, so you could try adding the tag (by the name when loaded with pydicom) here:

```
{
   "get": { 

         
         "skip": ["PixelData",
                  "RedPaletteColorLookupTableData",
                  "GreenPaletteColorLookupTableData",
                  "BluePaletteColorLookupTableData",
                  "VOILUTSequence"],

         "ids":{
            "entity":"PatientID",
            "item":"AccessionNumber"
         }

   },

   "put":{

          "actions":[

              {"action":"ADD" ,"field":"PatientIdentityRemoved","value": "Yes"},
              {"action":"KEEP","field":"PixelData"}
      ]
   }
}
```
Note that you can also do this by adding the field to the list for the `skip_fields` variable.
Give that a try and see if it maintains the tag in your data.

**Comment by vsoch** [comment]
Can you please give me enough to reproduce the error (meaning a deidentified file, and exactly the commands you are trying)? That way I can try to help.

**Comment by W7CAM** [comment]
I used the code provided deid-dicom-example.py (removing comments and a few other parts) to de-identify the files. This used the default recipe.

```python
from deid.dicom import get_files, replace_identifiers
from deid.utils import get_installdir
from deid.data import get_dataset
import os
df = list(get_files('~/test_folder/identified')) 
# This is the function to get identifiers
from deid.dicom import get_identifiers
ids = get_identifiers(dicom_files)
# Create a DeidRecipe
from deid.config import DeidRecipe
recipe = DeidRecipe()
cleaned_files = replace_identifiers(dicom_files=dicom_files,
                                    deid=recipe,
                                    ids=updated_ids,
                                    remove_private=True,
                                    output_folder='~/test_folder/deidentified/private_removed')
```
if I print the ids using
```python
print(ids)
```
the identifiers from 0051 are not there. 

And then used libdiff (another online example) to examine the differences between the identified and deidentified files

```python
import difflib

filename_ident = list(get_files('~/test_folder/identified'))[0] 
filename_deident = list(get_files('~/test_folder/deidentified'))[0] 

datasets = tuple([pydicom.dcmread(filename, force=True)
                  for filename in (filename_ident, filename_deident)])

rep = []
for dataset in datasets:
    lines = str(dataset).split("\n")
    lines = [line + "\n" for line in lines]  # add the newline to end
    rep.append(lines)

diff = difflib.Differ()
for line in diff.compare(rep[0], rep[1]):
    if line[0] == "-":
        print(line)
```

which showed that 0051 (and 0019) was removed from the file. This is simply because 0051 is not in pydicom's list (nor is it in the [NEMA](http://dicom.nema.org/Dicom/2013/output/chtml/part06/chapter_6.html) Dicom documentation...) This isn't so much a deid problem as it is the OEM inventing a new tag group. It looks like Nipy ran into this as well. 

Adding these to _dicom_dict.py in the pydicom library fixes the problem.
```
    0x00511008: ('CS', '1', "CSA Image Header Type", '', 'CSAImageHeaderType'), #noqa
    0x00511009: ('CS', '1', "CAS Image Header Version ??", '', 'CSAImageHeaderVersion'), #noqa
    0x0051100a: ('LO', '1', "Acquistion duration", '', 'TA'), #noqa
    0x0051100b: ('LO', '1', "Acquistiion Matrix Text", '', 'AcquistiionMatrixText'),
    0x0051100c: ('LO', '1', "Field of View cm", '', 'FoVcm'),
    0x0051100d: ('SH', '1', "Unknown1", '', 'Unknown1'),
    0x0051100e: ('LO', '1', "Unknown2", '', 'Unknown2'),
    0x0051100f: ('LO', '1', "Coil String", '', 'CoilString'),
    0x00511011: ('LO', '1', "PAT Mode Text", '', 'PATModeText'),
    0x00511012: ('SH', '1', "Unknown3", '', 'Unknown3'),
    0x00511013: ('SH', '1', "Positive PCS Directions", '', 'PositivePCSDirections'),
    0x00511015: ('SH', '1', "Unknown4", '', 'Unknown4'),
    0x00511016: ('SH', '1', "ReconstructionDetails", '', 'ReconstructionDetails'),
    0x00511017: ('SH', '1', "Unknown5", '', 'Unknown5'),
```

**Comment by vsoch** [comment]
Yes that makes sense, thank you for sharing this! But I need also the image(s) you are testing, cleaned if necessary,  in order to test this, reproduce, and then suggest/implement a fix. Thank you!

**Comment by vsoch** [comment]
hey @W7CAM I wanted to follow up on this- it probably should be the case that these tags are included in the pydicom dicom dict. Did you try a newer version and see if they are? If not I would open an issue there, and I can start to look into updating the version that deid uses. Thanks!

**Comment by Biggz1313** [comment]
I was just reading through this, and noticed the group in question is an odd numbered group.  Unless I'm mistaken, aren't odd numbered groups considered private tags?  If that's the case, there can't really be a standard to add these to the PyDICOM dictionary can they?  Anyone can choose to use group 0051 as their own tag and it would most likely be different across the board.  I could be off base on these assumptions because I've just recently really started to dig into Dicom standards, but just thought I'd speak up incase it helps.

**Comment by vsoch** [comment]
Your insight is very helpful @Biggz1313 ! In fact, we can look at where deid derives it's private tags, from pydicom. Here they are:

https://github.com/pydicom/pydicom/blob/master/pydicom/_private_dict.py

We might then select one that was missing:

```
0x00511008: ('CS', '1', "CSA Image Header Type", '', 'CSAImageHeaderType'), #13 
```
If you look for `CSA Image Header Type` in that file, you will find three instances of it, each associated with a particular model (e.g., Siemens MR Header). For this particular 0051 group, you can find it here:
https://github.com/pydicom/pydicom/blob/master/pydicom/_private_dict.py#L7760

I have only looked into this superficially, but it does appear to be a private tag, and is intended for a particular implementation.

> Private Creator Data Elements numbered (gggg,0010-00FF) (gggg is odd) shall be used to reserve a block of Elements with Group Number gggg for use by an individual implementer. The implementer shall insert an identification code in the first unused (unassigned) Element in this series to reserve a block of Private Elements. The VR of the private identification code shall be LO (Long String) and the VM shall be equal to 1.

And then what does pydicom do? (and deid honors) when we remove private tags ([this function](https://github.com/pydicom/pydicom/blob/master/pydicom/dataset.py#L1037)) we check if the `is_private` is True. This is determined based on looking if the group number is even:

```python
    @property
    def is_private(self):
        """Return True if the tag is private (has an odd group number)."""
        return self.group % 2 == 1
```
So @W7CAM - you will need to have extra care if you want to keep these tags. Have you tried setting `remove_private` to False, and then inspecting the data? See here --> https://github.com/pydicom/deid/blob/master/deid/dicom/header.py#L251
